### PR TITLE
Implementation of unit-testability for mapdql commands + template-based interface unit-test mocking solution for ClientContext/Thrift-Retry Operations, +Issue #135 Fix, +Issue #144 Proposal

### PR DESCRIPTION
**Pull request by ptumati**
- I refactored a portion of the strncmp()/strtok() stuff into a generalized command-resolution class that enables developers to simply provide a command to match, a functor to execute on match, a help message, and a minimum parameter count.  Example usage is below[1].  This moves the code away from the system which led to problems like that in Issue #135

- As part of the refactoring, I made it so command operations could be tested independently of having to actually issue thrift calls (although these calls can still be made) by abstracting away client context and thrift operations into a template meta class.  This class rolls up functions like get_all_roles() into static methods that can be mocked-away in unit tests.  An example of a cleaner, command-focused unit-test is below  [2]

- New commands should be introduced via CommandFunctors.h
- CommandResolutionChain.h has unit tests, and tests for mapdql commands were added to MapDQLCommandTest.cpp in the Tests/ directory

The intention is to replace the existing command processing behaviors over time.  However, new commands can be added, unit tested, and used without issue.

[1]  Note functors or lambdas can be used interchangeable.  Help messages are easily added as necessary.

`

      using Params = CommandResolutionChain<>::CommandTokenList;

      auto resolution_status =
          CommandResolutionChain<>(
              line, "\\copygeo", 3, CopyGeoCmd<>(context), "Usage: \\copygeo <serverGeoFileName> <tableName>")(
              "\\copy",
              3,
              [&](Params const& p) { copy_table(p[1].c_str() /* filepath */, p[2].c_str() /* table */, context); })(
              "\\ste", 2, [&](Params const& p) { set_table_epoch(context, p[1] /* table_details */); })(
              "\\gte", 2, [&](Params const& p) { get_table_epoch(context, p[1] /* table_details */); })(
              "\\export_dashboard", 2, [&](Params const& p) { export_dashboard(context, p[1] /* dash_details */); })(
              "\\import_dashboard", 2, [&](Params const& p) { import_dashboard(context, p[1] /*dash_details */); })(
              "\\role_list", 2, RoleListCmd<>(context), "Usage: \\role_list <userName>")(
              "\\roles", 1, RolesCmd<>(context))
              .is_resolved();

`

[2] Sample unit test with thrift-call interface mocking:

```
//
// \\copygeo Command Unit Test and Support Mockups
//

struct CopyGeoCommandMockupContext {
  std::string file_name;
  std::string table_name;
  bool import_geo_table_invoked = false;
};

struct CopyGeoCommandContextOpsPolicy {
  using ThriftServiceType = ThriftService;
  using ContextType = CopyGeoCommandMockupContext;
};

template <typename CONTEXT_OP_POLICY>
class CopyGeoCommandMockupContextOperations {
 public:
  using ThriftService = typename CONTEXT_OP_POLICY::ThriftServiceType;
  using ContextType = typename CONTEXT_OP_POLICY::ContextType;

  static void import_geo_table(ContextType& context) { context.import_geo_table_invoked = true; }
};

TEST(MapDQLTest, CopyGeoCommandTest) {
  using Params = CommandResolutionChain<>::CommandTokenList;
  using UnitTestCopyGeoCmd = CopyGeoCmd<CopyGeoCommandContextOpsPolicy, CopyGeoCommandMockupContextOperations>;
  CopyGeoCommandMockupContext unit_test_context;

  auto resolution = CommandResolutionChain<>(
                        "\\copygeo filename.csv tablename", "\\copygeo", 3, UnitTestCopyGeoCmd(unit_test_context))
                        .is_resolved();

  EXPECT_TRUE(unit_test_context.import_geo_table_invoked);
  EXPECT_TRUE(resolution);
}

```

### CODE REVIEW

**Comment by ptumati** [comment]
Cool.  This front-end refactoring was related to #143 in that I saw @billmaimone request unit tests on a submission involving wildcard matches (the implementation submitted prior to my regex submission.)  I thought it made sense to formalize command creation and unit-testing, so I could add unit-tests for #143!  I would refactor process_backslash_commands() to use the facilities I added in this pull request to unit test those submissions.

**Comment by shtilman** [comment]
Yep, sounds great. Thanks

**Comment by ptumati** [comment]
I'll bring this pull request up to speed and resubmit.  The new changes on master would break (some) things.  Conceptually, with regards to the proposal, not much has changed.

**Comment by ptumati** [comment]
Rebased and squashed; updated.  Applicable to master once more.

**Comment by ptumati** [comment]
Updated.  @andrewseidl, I noticed your license command work.  When I rebased I added your commands to the system created in this commit.  I've just moved the code from the stub command into a lambda for now, but commands can now be unit tested and the C functions for the license operations could be moved into statics for ContextOperations, see CommandFunctors.h and ContextOperations within that class.

I'll move a few commands in and add the regex unit tests after this pull gets accepted.

File: SQLFrontend/ThriftWithRetry.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
+#ifndef THRIFTWITHRETRY_H
+#define THRIFTWITHRETRY_H
+
```
[the remark this answers is not in the document]
Comment (shtilman) [reply SQLFrontend/ThriftWithRetry.h]: Sorry, this should have been
`#include <cmath>`

File: SQLFrontend/ClientContext.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+#include "gen-cpp/mapd_types.h"
+
+using namespace ::apache::thrift;
+using namespace ::apache::thrift::protocol;
+using namespace ::apache::thrift::transport;
+
+static constexpr char const* INVALID_SESSION_ID("");
+static constexpr char const* MAPD_ROOT_USER("mapd");
+static constexpr char const* MAPD_DEFAULT_ROOT_USER_ROLE("mapd_default_suser_role");
```
Comment (shtilman) [inline SQLFrontend/ClientContext.h]: Why not std::string?

Comment (ptumati) [reply SQLFrontend/ClientContext.h]: They can be std::string if it works.  

When I did the first pass conversion,  I wasn't entirely sure whether any of the C-style functions were dependent on these definitions from elsewhere at first glance.  I figured anywhere that took a std::string would work automatically and the C functions would still work.  If the std::string works without issue or makes the code better, I can make the change.

File: SQLFrontend/ThriftWithRetry.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ptumati) [reply SQLFrontend/ThriftWithRetry.h]: OK.  Will correct.

File: SQLFrontend/ClientContext.h (hunk 2 above)
Comment (ptumati) [reply SQLFrontend/ClientContext.h]: Looks like INVALID_SESSION_ID only compared against a string in ClientContext.  MAPD_DEFAULT_ROOT_USER_ROLE not really found in SQLFrontend/* and MAPD_ROOT_USER used once in some informational privileges message.  They're probably safe for conversion.  I'll change them now.

**Comment by ptumati** [comment]
Everything corrected.

**Comment by ptumati** [comment]
Rebased again.  Assuming this is fair game for a merge since the 3.4 release seems to have gone out.  Let me know if there are some issues.  I'd like to convert more commands and add the requested unit tests without having to navigate too many conflicts.

**Comment by shtilman** [comment]
@dwayneberry could you glance at the refactored mapdql before @andrewseidl pulls this in? Anything else that needs to be adjusted?

**Comment by ptumati** [comment]
Closing the old pull request.

**Pull request closed without merging**